Sources/NIOPosix/PendingDatagramWritesManager.swift

```diff
@@ -515,14 +566,31 @@ final class PendingDatagramWritesManager: PendingWritesManager {
     ///
     /// - parameters:
     ///     - scalarWriteOperation: An operation that writes a single, contiguous array of bytes (usually `sendto`).
-    private func triggerScalarBufferWrite(scalarWriteOperation: (UnsafeRawBufferPointer, UnsafePointer<sockaddr>, socklen_t, AddressedEnvelope<ByteBuffer>.Metadata?) throws -> IOResult<Int>) rethrows -> OneWriteOperationResult {
+    private func triggerScalarBufferWrite(scalarWriteOperation: (UnsafeRawBufferPointer, UnsafePointer<sockaddr>?, socklen_t, AddressedEnvelope<ByteBuffer>.Metadata?) throws -> IOResult<Int>) rethrows -> OneWriteOperationResult {
         assert(self.state.isFlushPending && self.isOpen && !self.state.isEmpty,
                "illegal state for scalar datagram write operation: flushPending: \(self.state.isFlushPending), isOpen: \(self.isOpen), empty: \(self.state.isEmpty)")
         let pending = self.state.nextWrite!
         do {
-            let writeResult = try pending.address.withSockAddr { (addrPtr, addrSize) in
-                try pending.data.withUnsafeReadableBytes {
-                    try scalarWriteOperation($0, addrPtr, socklen_t(addrSize), pending.metadata)
+            let writeResult: IOResult<Int>
+
+            if let address = pending.address {
+                assert(!self.state.isConnected, "Pending write with address on connected socket.")
```

Comment:
So the problem with reflecting that semantics is the issue of `AddressedEnvelope.Metadata`. This is still meaningful to apply even in connected mode, but it's not accessible to users anymore if they connect. I don't think that's an intended outcome. Supporting both and just fixing it up allows users to wish to to keep sending the metadata but still gives access to a fast path. Does that make sense?

I don't think this conflicts with late-binding the address though: the `DatagramChannel` just needs to communicate the metadata to the PDWM, at which point it can safely drop the address before storing into the `PendingDatagramWrite`. Does that make sense?

Reply:
Also you're right about the vector case, my bad, I just missed it.

Reply:
Your statements make sense. I thought, however, we wanted to offer APIs in terms of `ByteBuffer` and if we offer such APIs we should guide/force users to use the appropriate one.

Right, the point about the metadata is interesting and I came across it when I was making changes to PDWM. The additional flags passed to the `sendmsg` are a function of not just the `AddressedEnvelope.metadada` but also the `AddressedEnvelope.address.protocol`. I played around with having this property lifted into the `PendingDatagramWrite` so that it can be used even when `AddressedEnvelope.address` is `nil`.

I think if we want to continue to support these additional bits of metadata along with the write, which I completely expect that we do, we have (at least) two options:

1. `DatagramChannel.writeToSocket(_:)` accepts just an `AddressedEnvelope` and the implementation (either channel, PDWM, or both) discards the address along the way. This is an implicit API which might be harder to reason about.
2. Expose a more  explicit API that takes `ByteBuffer` and expect users to use it when connected. This API would need some way of also passing in the metadata that is appropriate, i.e. the congestion control mechanism currently composed of `AddressedEnvelope.metadata` and `AddressedEnvelope.address.protocol`, without passing the the metadata that is inappropriate, i.e. the destination address.

Reply:
Nope, my preference is that `PendingDatagramWrite` has a `nil` address when we don't want it, and then just restructure how we construct the metadata to use the connected address from `DatagramChannel` to populate the `protocol` for the metadata early.

As to the broader question of forcing folks to use the "right" API, I think unless we're willing to change the write to be "ByteBuffer + optional metadata" (which I think is not great) then it's probably fine to just support `AddressedEnvelope` for now in cases where folks want to send ECN.

Reply:
OK, I relaxed support to allow for `AddressedEnvelope` in writes to a connected `DatagramChannel` in eb580d69c. It validates that the address in the envelope matches that of the connected socket.